### PR TITLE
cdc: remove downstreams correctly when uninitialized

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What is changed and how it works?

https://github.com/tikv/tikv/pull/11896#discussion_r834951841 introduces a critical bug for CDC, which can cause data lost or downstreams blocking forever. Here is how to trigger the bug:
1. multiple downstreams subscribe one region
2. cancel one downstream before the region delegate has been initialized

Then, `Delegate::stop_observing` will be called incorrectly and all change events will be lost.

This PR does
1. correct the bug
2. add a log for `stop_observing`
3. complete unit tests for `Delegate::subscribe` and `Delegate::unsubscribe`.

```commit-message
cdc: remove downstreams correctly when uninitialized
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/delegate.rs [hunk 1]
```diff
@@ -760,7 +761,7 @@ impl Delegate {
         let downstreams = self.downstreams_mut();
         if let Some(index) = downstreams.iter().position(|x| x.id == id) {
             let downstream = downstreams.swap_remove(index);
-            if self.downstreams.is_empty() {
+            if downstreams.is_empty() {
                 self.stop_observing();
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:765]: Could you add some comments about when and why we need to `stop_observing`?

**Review by overvenus — APPROVED** [review]
Rest LGTM

**Review by hi-rustin — APPROVED** [review]

**Pull request merged**